### ISSUE TITLE
Analysis of the Mining Process

### ISSUE THREAD

**Issue by WhoSoup** [pegnet/pegnet#83]
First of all, the LXRHash cascading effect seems to be doing its job fairly well. I ran an analysis on the results of the mining algorithm and determined that the resulting hashes are uniformly distributed:

https://i.imgur.com/0X8Zn4L.png

This data contains 1,907,966 hashes, calculated with `LXRHash(LXRHash("foo") | nonce)` using the miner nonce format. For each hash, the "difficulty" was calculated and sorted into a bucket based on their percentile, spanning the entire range of Uint64 divided into 1000 buckets. Bucket 0 is `0.0%` to `0.1%` and bucket 999 is `99.9%`, etc.

At 1.9m hashes and 1000 buckets, we would expect the average bucket to hold 1,908 items. This turns out to be true as seen in the histogram. The normal distribution of the histogram is a strong indicator that the underlying data is random.

Conclusion: The results of hashing data with LXRHash are fairly random.

Now, this has some implications to the efficacy of mining in general. If you know the the time period (9 minutes) and the Hashrate (depends on setup), it is possible to calculate the likelihood that you will reach a higher difficulty than the one you attained.

I started out my analysis by just letting miners run and see what kind of difficulties they reach and how close to the maximum `(2^64)-1` (18446744073709551615) they get. Turns out this happens rather fast. In most mining runs, the miners hit an early maximum within the first one or two minutes and occasionally improving on that maximum down the road. These results are deterministic depending on the base OPR Hash, but will vary depending on how fast your CPU can mine.

Using the base of `LXRHash("foo")` for example, I reached the difficulty `18446742286272833685` in the first minute (minute 0), improving to `18446743757683897496` in the tenth minute (minute 9). Since these numbers are hard to parse, I'm going to use a different format, which is `MaxUint64 - difficulty` in e-notation. The difficulty `18446742286272833685` becomes `1.7e12`. This is the distance to the maximum, so **lower is better**.

To give these numbers some reference, `1e13` is at [~99.99995%](https://www.wolframalpha.com/input/?i=(18446744073709551615-+1e13)+%2F+18446744073709551615) (4 nines) of the maximum. `1e12` is [~99.999995%](https://www.wolframalpha.com/input/?i=(18446744073709551615-+1e12)+%2F+18446744073709551615) (5 nines) of the maximum. `1e11` is [~99.9999995%](https://www.wolframalpha.com/input/?i=(18446744073709551615-+1e11)+%2F+18446744073709551615) (6 nines).

| base | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |
|---|---|---|---|---|---|---|---|---|---|---|---|
|foo|1.7e12|-|-|-|-|-|-|-|-|3.2e11|-|
|bar|5.3e12|2.9e12|-|-|-|-|-|-|-|-|-|
|boo|7.7e12|-|4.5e12|-|-|1.1e12|-|-|-|-|-|
|car|1.1e13|-|-|-|-|3.8e12|-|-|-|3.0e12|-|
|zoo|8.6e12|-|5.1e12|-|-|-|-|-|-|-|-|
|dar|1.2e13|-|1.0e12|-|-|-|-|-|-|-|-|

Under the assumption that LXRHash is random, we can now calculate the likelihood of a single hash reaching a higher difficulty: `P = difficulty / MaxUint64` or the chance that a hash has the same or lower difficulty: `P2 = (MaxUint64 - difficulty) / MaxUint64`. My hashrate on a 9-year-old rig is: 13,374 hashes/second or to make it easier, 800k per minute.

Let's assume that we reached `1e13` during the first minute. We have eight more minutes left, or 6.4 million hashes. The likelihood of 6.4 million hashes all being less or equal to `1e13` is `P2 ^ 6.4e6` or `((2^64 - 1 - 1e13) / (2 ^ 64 - 1))^(6.4e6)`, which [turns out to be](https://www.wolframalpha.com/input/?i=((2^64+-+1+-+1e13)+%2F+(2+^+64+-+1))^(6.4e6)) 3.1%. That means there's a 97% likelihood one of the hashes will have a higher difficulty.

When we look at `1e12` however, the outcome of `P2 ^ 6.4e6` is 70.6%. That means there is only a 30% likelihood one of the hashes will have a higher difficulty. At `1e11`, the outcome is 96.6% giving us only a 3.4% likelihood that one of the hashes will improve. And this is, of course, for the result after the first minute.

Let's take the table above and include the data of how likely we are to get a better difficulty in the remaining time:

| base | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |
|---|---|---|---|---|---|---|---|---|---|---|
|foo|1.7e12<br>45%|-|-|-|-|-|-|-|-|3.2e11|
|bar|5.3e12<br>84%|2.9e12<br>58%|-|-|-|-|-|-|-|-|
|boo|7.7e12<br>93%|-|4.5e12<br>69%|-|-|1.1e12<br>13%|-|-|-|-|
|car|1.1e13<br>98%|-|-|-|-|3.8e12<br>39%|-|-|-|3.0e12|
|zoo|8.6e12<br>95%|-|5.1e12<br>73%|-|-|-|-|-|-|-|
|dar|1.2e13<br>98%|-|1.0e12<br>23%|-|-|-|-|-|-|-|

The sample size here is fairly low but you can see that the ones with high percentages end up getting replacements fairly quickly. An outlier here is "zoo" which had a 1-in-4 likelihood of not getting one. 

A higher hashrate means these probabilities increase. Let's assume we got a result of 1e12 and we have 8 minutes left to mine.

|Hash/m | Res |
| --- | --- |
| 100k | 4% | 
| 200k | 8% | 
| 200k | 16% | 
| 800k | 29% | 
| 1.6m | 50% | 
| 3.2m | 75% | 
| 6.4m | 94% | 

Using this process, it is possible to estimate whether or not a miner should abandon their search. This means a single miner may only take a fraction of the total mining time, say three or four minutes. Further, running multiple miners means you can share the highest difficulties reached across the mining pool, letting you keep a "Top X" list of difficulties. You can keep mining as long as the lowest difficulty in the list has a likelihood within a certain treshold.

**Comment by PaulBernier** [comment]
Really nice analysis, thank you! About the first part, glad to see the distribution of the output is reasonable. But to ever feel safe I'd probably need a crypt-analyse of the hash function against pre-image & other attacks. Unfortunately I don't have those skills :(

About there rest, I am taking good notes of your findings hehe ;)

**Comment by PaulBernier** [comment]
Thanks again for your analysis. Do you have any objection of closing this issue as I mostly see it as an informational post? Or do you expect more discussion/feedback around it?

**Comment by WhoSoup** [comment]
> Or do you expect more discussion/feedback around it?

When I wrote it, I did it with the hope that we could bake the option of partial-time-mining into the miner itself to give miners the option of saving electricity. I'm not sure if that's too much to implement at this stage of development, so I'm okay with closing the issue and coming back to it at a later point once pegnet is up and running.
